File: ios/PlaytomicFoundation/executor/ExecutorQueue.swift
//
//  ExecutorQueue.swift
//  PlaytomicFoundation
//
//  Created by Angel Luis Garcia on 04/09/2020.
//  Copyright © 2020 Playtomic. All rights reserved.
//

import Foundation

public class ExecutorQueue {
    public var taskCount: Int { pendingTasks.count }
    public var isRunning: Observable<Bool> { _isRunning }
    public var isEmpty: Bool { pendingTasks.isEmpty }
    private let _isRunning = MutableObservable<Bool>(value: false)
    private var pendingTasks: [Task] = []

    public init() { }

    @discardableResult
    public func add<T>(_ promiseProvider: @escaping (() -> Promise<T>)) -> Promise<T> {
        let promiseTask = PromiseTask(promiseProvider: promiseProvider)
        add(task: promiseTask)
        return promiseTask.pendingPromise
    }

    private func add(task: Task) {
        pendingTasks.append(task)
        if isRunning.value == false {
            processNext()
        }
    }

    private func processNext() {
        if !pendingTasks.isEmpty {
            _isRunning.distinctValue = true
            pendingTasks.removeFirst().execute().always(processNext)
        } else {
            _isRunning.distinctValue = false
        }
    }
}

private protocol Task {
    func execute() -> Promise<Void>
}

private class PromiseTask<T>: Task {
    let promiseProvider: () -> Promise<T>
    let pendingPromise: PendingPromise<T>

    init(promiseProvider: @escaping (() -> Promise<T>)) {
        self.promiseProvider = promiseProvider
        self.pendingPromise = PendingPromise()
    }

    func execute() -> Promise<Void> {
        promiseProvider()
            .then(pendingPromise.fulfill)
            .catchError(pendingPromise.reject)
            .then(map: { _ in () })
            .fulfillOnError(value: ())
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/executor/ExecutorQueue.kt
package com.playtomic.foundation.executor

import com.playtomic.foundation.observable.MutableObservable
import com.playtomic.foundation.observable.Observable
import com.playtomic.foundation.promise.PendingPromise
import com.playtomic.foundation.promise.Promise

class ExecutorQueue {
    val taskCount: Int get() = pendingTasks.count()
    val isRunning: Observable<Boolean>
        get() = _isRunning
    val isEmpty: Boolean get() = pendingTasks.isEmpty()
    private val _isRunning = MutableObservable<Boolean>(value = false)
    private var pendingTasks = mutableListOf<Task>()

    fun <T> add(promiseProvider: (() -> Promise<T>)): Promise<T> {
        val promiseTask = PromiseTask(promiseProvider = promiseProvider)
        add(task = promiseTask)
        return promiseTask.pendingPromise
    }

    private fun add(task: Task) {
        pendingTasks.add(task)
        if (isRunning.value == false) {
            processNext()
        }
    }

    private fun processNext() {
        if (!pendingTasks.isEmpty()) {
            _isRunning.distinctValue = true
            pendingTasks.removeAt(0).execute().always(this::processNext)
        } else {
            _isRunning.distinctValue = false
        }
    }
}

private interface Task {
    fun execute(): Promise<Unit>
}

private class PromiseTask<T>(
    val promiseProvider: (() -> Promise<T>)
) : Task {
    val pendingPromise: PendingPromise<T> = PendingPromise()

    override fun execute(): Promise<Unit> =
        promiseProvider()
            .then(pendingPromise::fulfill)
            .catchError(pendingPromise::reject)
            .then(map = { _ -> Unit })
            .fulfillOnError(value = Unit)
}
